File: Tests/DoNotTypeCoreTests/SpeechActivityTests.swift
import Foundation
import XCTest

@testable import DoNotTypeCore

/// Silero is asserted against recordings rather than mocked probabilities. These fixtures are
/// shared with the other clients, so a model/runtime/normalisation mistake cannot pass as long as
/// the post-processing state machine happens to be correct.
final class SpeechActivityTests: XCTestCase {
    private func fixture(_ path: String) throws -> Data {
        var directory = URL(fileURLWithPath: #filePath)
        for _ in 0..<6 {
            directory.deleteLastPathComponent()
            let candidate = directory.appendingPathComponent(path)
            if FileManager.default.fileExists(atPath: candidate.path) {
                return try Data(contentsOf: candidate)
            }
        }
        throw XCTSkip("fixture \(path) not found")
    }

    private func silence(_ name: String) throws -> Data {
        try fixture("eval/audio/silence/\(name).wav")
    }

    // MARK: - Nothing that is not speech gets through

    func testNothingWithoutSpeechIsEverSent() throws {
        for name in ["digital-silence", "room-tone", "steady-noise", "hum", "too-short"] {
            let reading = try SpeechActivity.measure(wav: silence(name))
            XCTAssertFalse(
                reading.hasSpeech,
                "\(name) would have been sent to a model — \(reading.summary)")
            XCTAssertEqual(reading.speechMilliseconds, 0, reading.summary)
            XCTAssertLessThan(reading.maximumProbability, 0.2, reading.summary)
        }
    }

    func testKeyboardAndMouseClicksAreNotSentences() throws {
        for name in ["click", "mouse-click-quiet-room"] {
            let reading = try SpeechActivity.measure(wav: silence(name))
            XCTAssertFalse(reading.hasSpeech, "\(name) — \(reading.summary)")
            XCTAssertLessThan(reading.maximumProbability, 0.2, reading.summary)
        }
    }

    // MARK: - Everything that is speech gets through

    func testAOneWordAnswerIsStillASentence() throws {
        let reading = try SpeechActivity.measure(wav: fixture("eval/audio/short-word.wav"))
        XCTAssertTrue(reading.hasSpeech, reading.summary)
        XCTAssertGreaterThan(reading.maximumProbability, 0.9, reading.summary)
        XCTAssertGreaterThanOrEqual(
            reading.speechMilliseconds, SpeechActivity.minimumSpeechMilliseconds,
            reading.summary)
    }

    func testRealSpeechIsAlwaysSent() throws {
        let reading = try SpeechActivity.measure(pcm: speechPCM())
        XCTAssertTrue(reading.hasSpeech, reading.summary)
        XCTAssertGreaterThan(reading.maximumProbability, 0.9, reading.summary)
    }

    /// Level-independent behaviour is a model property now, rather than something approximated
    /// by defining a threshold relative to the recording itself.
    func testQuietSpeechIsStillSpeech() throws {
        let speech = try speechPCM()
        for attenuation in [12, 20, 32, 40, 46, 52] {
            let reading = try SpeechActivity.measure(
                pcm: attenuated(speech, byDecibels: Double(attenuation)))
            XCTAssertTrue(
                reading.hasSpeech,
                "speech at −\(attenuation) dB would have been dropped — \(reading.summary)")
        }
    }

    /// The production failure this replacement is for. Aggressive microphone gain control can
    /// leave no quiet tenth of a recording: the old detector called the quieter parts of the voice
    /// its noise floor and rejected the whole utterance. Companding a real speech fixture emulates
    /// that narrow dynamic range while preserving its temporal and spectral shape.
    func testContinuousGainControlledSpeechNeedsNoQuietFloor() throws {
        let reading = try SpeechActivity.measure(pcm: companded(try speechPCM()))
        XCTAssertTrue(reading.hasSpeech, reading.summary)
        XCTAssertGreaterThan(reading.maximumProbability, 0.9, reading.summary)
    }

    // MARK: - Shape and diagnostics

    func testAnEmptyRecordingIsNotSpeech() throws {
        XCTAssertFalse(try SpeechActivity.measure(pcm: Data()).hasSpeech)
        XCTAssertThrowsError(try SpeechActivity.measure(wav: Data()))
    }

    func testAFragmentShorterThanTheMinimumIsNotSpeech() throws {
        let tenMilliseconds = Data(count: 16_000 / 100 * 2)
        XCTAssertFalse(try SpeechActivity.measure(pcm: tenMilliseconds).hasSpeech)
    }

    func testSummaryNamesTheDetectorAndCarriesProbabilities() throws {
        let summary = try SpeechActivity.measure(pcm: speechPCM()).summary
        XCTAssertTrue(summary.contains("silero"), summary)
        XCTAssertTrue(summary.contains("speech="), summary)
        XCTAssertTrue(summary.contains("max="), summary)
        XCTAssertTrue(summary.contains("mean="), summary)
    }

    // MARK: - Helpers

    private func speechPCM() throws -> Data {
        let wav = try fixture("eval/audio/formats/speech.wav")
        return try XCTUnwrap(AudioChunker.pcmBody(of: wav))
    }

    private func attenuated(_ pcm: Data, byDecibels decibels: Double) -> Data {
        let factor = pow(10.0, -decibels / 20.0)
        return mapSamples(pcm) { Double($0) * factor }
    }

    private func companded(_ pcm: Data) -> Data {
        mapSamples(pcm) { sample in
            let normalised = abs(Double(sample)) / 32_768
            let magnitude = pow(normalised, 0.2) * 12_000
            return sample < 0 ? -magnitude : magnitude
        }
    }

    private func mapSamples(_ pcm: Data, _ transform: (Int16) -> Double) -> Data {
        var out = Data(capacity: pcm.count)
        pcm.withUnsafeBytes { raw in
            for index in 0..<(pcm.count / 2) {
                let source = Int16(littleEndian: raw.loadUnaligned(
                    fromByteOffset: index * 2, as: Int16.self))
                let transformed = transform(source).rounded()
                let mapped = Int16(min(32_767, max(-32_768, transformed)))
                withUnsafeBytes(of: mapped.littleEndian) { out.append(contentsOf: $0) }
            }
        }
        return out
    }
}

File: windows/DoNotType.Core.Tests/SpeechActivityTests.cs
using DoNotType.Core;
using Xunit;

namespace DoNotType.Core.Tests;

/// <summary>
/// Silero itself is asserted against the shared recordings. Mocking probabilities would only test
/// our small state machine and miss a wrong tensor shape, sample normalisation or packaged model.
/// </summary>
public sealed class SpeechActivityTests
{
    private static byte[] Fixture(string relative)
    {
        var directory = AppContext.BaseDirectory;
        for (var depth = 0; depth < 8; depth++)
        {
            var candidate = Path.Combine(directory, relative);
            if (File.Exists(candidate)) return File.ReadAllBytes(candidate);
            directory = Path.Combine(directory, "..");
        }
        throw new FileNotFoundException($"fixture {relative} not found from {AppContext.BaseDirectory}");
    }

    private static byte[] Silence(string name) =>
        Fixture(Path.Combine("eval", "audio", "silence", $"{name}.wav"));

    private static byte[] SpeechPcm() =>
        AudioChunker.PcmBody(Fixture(Path.Combine("eval", "audio", "formats", "speech.wav")))!;

    // ---- Nothing that is not speech gets through --------------------------------------------

    [Theory]
    [InlineData("digital-silence")]
    [InlineData("room-tone")]
    [InlineData("steady-noise")]
    [InlineData("hum")]
    [InlineData("too-short")]
    public void NothingWithoutSpeechIsEverSent(string name)
    {
        var reading = SpeechActivity.MeasureWav(Silence(name));
        Assert.False(reading.HasSpeech, $"{name} would have been sent — {reading.Summary}");
        Assert.Equal(0, reading.SpeechMilliseconds);
        Assert.True(reading.MaximumProbability < 0.2, reading.Summary);
    }

    [Theory]
    [InlineData("click")]
    [InlineData("mouse-click-quiet-room")]
    public void KeyboardAndMouseClicksAreNotSentences(string name)
    {
        var reading = SpeechActivity.MeasureWav(Silence(name));
        Assert.False(reading.HasSpeech, $"{name} — {reading.Summary}");
        Assert.True(reading.MaximumProbability < 0.2, reading.Summary);
    }

    // ---- Everything that is speech gets through --------------------------------------------

    [Fact]
    public void AOneWordAnswerIsStillASentence()
    {
        var reading = SpeechActivity.MeasureWav(
            Fixture(Path.Combine("eval", "audio", "short-word.wav")));
        Assert.True(reading.HasSpeech, reading.Summary);
        Assert.True(reading.MaximumProbability > 0.9, reading.Summary);
        Assert.True(
            reading.SpeechMilliseconds >= SpeechActivity.MinimumSpeechMilliseconds,
            reading.Summary);
    }

    [Fact]
    public void RealSpeechIsAlwaysSent()
    {
        var reading = SpeechActivity.Measure(SpeechPcm());
        Assert.True(reading.HasSpeech, reading.Summary);
        Assert.True(reading.MaximumProbability > 0.9, reading.Summary);
    }

    [Theory]
    [InlineData(12)]
    [InlineData(20)]
    [InlineData(32)]
    [InlineData(40)]
    [InlineData(46)]
    [InlineData(52)]
    public void QuietSpeechIsStillSpeech(int attenuation)
    {
        var reading = SpeechActivity.Measure(Attenuated(SpeechPcm(), attenuation));
        Assert.True(reading.HasSpeech, $"speech at -{attenuation} dB — {reading.Summary}");
    }

    /// <summary>
    /// Emulates the narrow dynamic range of the real continuous-speech recordings the old
    /// recording-relative noise floor rejected.
    /// </summary>
    [Fact]
    public void ContinuousGainControlledSpeechNeedsNoQuietFloor()
    {
        var reading = SpeechActivity.Measure(Companded(SpeechPcm()));
        Assert.True(reading.HasSpeech, reading.Summary);
        Assert.True(reading.MaximumProbability > 0.9, reading.Summary);
    }

    // ---- Shape and diagnostics ---------------------------------------------------------------

    [Fact]
    public void EmptyAndInvalidRecordingsAreNotSpeech()
    {
        Assert.False(SpeechActivity.Measure([]).HasSpeech);
        Assert.Throws<ArgumentException>(() => SpeechActivity.MeasureWav([]));
    }

    [Fact]
    public void AFragmentShorterThanTheMinimumIsNotSpeech()
    {
        Assert.False(SpeechActivity.Measure(new byte[16_000 / 100 * 2]).HasSpeech);
    }

    [Fact]
    public void SummaryNamesTheDetectorAndCarriesProbabilities()
    {
        var summary = SpeechActivity.Measure(SpeechPcm()).Summary;
        Assert.Contains("silero", summary);
        Assert.Contains("speech=", summary);
        Assert.Contains("max=", summary);
        Assert.Contains("mean=", summary);
    }

    private static byte[] Attenuated(byte[] pcm, double decibels)
    {
        var factor = Math.Pow(10, -decibels / 20);
        return MapSamples(pcm, sample => sample * factor);
    }

    private static byte[] Companded(byte[] pcm) => MapSamples(pcm, sample =>
    {
        var magnitude = Math.Pow(Math.Abs(sample) / 32_768d, 0.2) * 12_000;
        return sample < 0 ? -magnitude : magnitude;
    });

    private static byte[] MapSamples(byte[] pcm, Func<short, double> transform)
    {
        var output = new byte[pcm.Length];
        for (var index = 0; index < pcm.Length / 2; index++)
        {
            var sample = BitConverter.ToInt16(pcm, index * 2);
            var mapped = (short)Math.Clamp(Math.Round(transform(sample)), short.MinValue, short.MaxValue);
            BitConverter.TryWriteBytes(output.AsSpan(index * 2, 2), mapped);
        }
        return output;
    }
}
